text marking - fade in animation

vaqueraoscar0 opened the pull request:
## Description Issue #92
Added a smooth animation to the text markings

## image/video 

https://github.com/OpenCovenant/quill/assets/57508905/cb7a9dcb-9d24-42bb-949c-ffda0ffb305d

AndiBraimllari commented:
That looks awesome! Well done. The code looks good from a first look, I guess it should be extended not just to typo markings but to loanword and stylistic markings as well (see line 301 in the same file), but that's trivially doable.

I am trying to think from a user's POV, and I imagine a case in which there is a lot of text in the editor with many markings, and the user types one more word (which will marked), unfortunately from what I can tell all markings (including the previous ones and the new one) will have this fade-in effect, but I believe only the "latest" one should have it (in order to signal to the user that these new markings i.e. the ones with the fade-in effect are the ones that have just been generated, and everything else is as it was). I believe adding such a thing would require some sort of "state" (i.e. knowing which is the latest markings or not) which may or may not

vaqueraoscar0 commented:
@AndiBraimllari Right, making changes to the .css file will not be enough to achieve this, are you okay with me tinkering with the other files? the problem here is that the elements are being created outside of the html file, since thats the case i need to be able to write some logic within home component or the text-marking am not too sure yet, i have some ideas but i need to test them first.

AndiBraimllari commented:
You're exactly right, home component or the text-marking file is where I'd also focus on, feel free to tinker with these files.
I think it is pretty interesting to design such a change (how do we calculate the state changes? how do we store the "state" to begin with? what should we include in the state? what data structure do we need?). I think tests here are pretty critical (more so from a user's POV, so we know what we want, rather than to actually test the code).
Feel free to ask any questions that come to mind!

vaqueraoscar0 commented:
@AndiBraimllari Great! I'll go ahead and start working on this then, and if I have any questions regarding the structure of the code or anything else I'll go ahead and let you know.

AndiBraimllari commented:
Awesome, some ideas that are coming to mind.
- the state can be Markovian (only the present matters for the future), i.e. we can only keep the last state
- perhaps we can store the inner HTML of the editor, then iterate each paragraph (skip the ones that are identical to the current one), and the iterate the children (or markings, or both) of the paragraphs that are different,   
- cleanup has to be done (remove any CSS classes that aren't relevant, cleanup the state etc)
- input of the function that determines which markings should be
  - old editor
  - new editor
  - current markings
- and of course the output should be a list of markings that should have the fade-in effect
- this function should graciously fail if it is canceled right away (i.e. any styling that is being done will have to be interrupted if the user clears the editor, some other marking occurs etc).

These are some design thoughts that came to mind while thinking about this. Wrote them in case they might be helpful to you, but of course feel free to implement it how you best see fit.

vaqueraoscar0 commented:
@AndiBraimllari Right! i've applied a few concept including some of the methods you've mentioned, i have come up with a solution but am still not quite happy with it. I might try a few other things before, and do some more testing. 

https://github.com/OpenCovenant/quill/assets/57508905/39d392e2-9bd8-4dca-b3a3-684fb58b99a4

AndiBraimllari commented:
Looks very good!

What do you think about this behavior? I feel it is necessary in the cases of text with many markings.

Keep it up, and of course if you'd like to discuss technical details I'm more than happy to do so, simply let me know.

vaqueraoscar0 commented:
Yes, i agree with you. It makes for a better user experience and will make the service seem more premium. And thank you, i will if i need further guidance.

vaqueraoscar0 commented:
## New Feature Description
You can now expect to get individual animation for any node that has been added, updated, or removed, this includes any node that could be potentially added at any specific index/place. 

## Issues 
As of now, i have yet to run into any issues. I initially wanted to include all of the logic within the home.compoent.ts file however, some delay was occurring which made it difficult to implement within that specific file. **Solution:** I applied all of the logic within the text-marking.ts which in the end turn out to be the better solution. The code is by no means perfect and any feedback is welcomed. 

## Demonstration Video

https://github.com/OpenCovenant/quill/assets/57508905/77e7bdec-21e4-4582-80f0-045444b8d4d5

## Additional Comments
Feel free to make any suggestion regarding the code structure.

@AndiBraimllari

AndiBraimllari commented:
Hey, I'll write more specific feedback soon. In the meanwhile, can you write some tests for this? Also, you can try running `npm run prettier:fix` to format it a bit. You can also add function return types to some methods.

AndiBraimllari commented on src/app/text-marking/text-marking.ts:
```diff
@@ -99,10 +111,132 @@ export function markText(
+    highlightedMarkingWords.push(newTextContent)
+
+    // Check if there are any previous markings
+    if(lastHighlightedMarkingWords.length === 0){
+        // If no previous markings, apply animation to the new node
+        newNode.style.animation = 'fillBackground 500ms linear forwards';
+    }else{
+        // Check various scenarios based on changes in markings
+        if(lastHighlightedMarkingWords.length < markingCount){
+            // If fewer markings than expected, apply animation to the new node
+            newNode.style.animation = 'fillBackground 500ms linear forwards';
+        }else if(lastHighlightedMarkingWords[markingCount] === newTextContent){
+            // If the current marking is the same, remove animation
+            newNode.style.animation = '';
+        }else if(lastHighlightedMarkingWords[markingCount] !== newTextContent){
+            // Handle scenarios for additions, removals, and modifications
+            if(lastHighlightedMarkingWords[markingCount + 1] === newTextContent){
+                newNode.style.animation = '';
+                // Remove the item if it was changed back to the previous content
+                removeItemAtIndex(markingCount)
+            }else if(markingCount === 0 && lastHighlightedMarkingWords[markingCount + 1] !== newTextContent){
+                // If a new first node was added, insert it and apply animation
+                addItemAtIndex(0, newTextContent)
+                newNode.style.animation = 'fillBackground 500ms linear forwards';
+            }else if(lastHighlightedMarkingWords[markingCount - 1] !== newTextContent || lastHighlightedMarkingWords[markingCount + 1] !== newTextContent){
+                // If the node was modified or inserted somewhere in the middle, apply animation
+                addItemAtIndex(markingCount, newTextContent)
+                newNode.style.animation = 'fillBackground 500ms linear forwards';
+            }else{
+                newNode.style.animation = 'fillBackground 500ms linear forwards';
+            }
+
+        }else{
+            newNode.style.animation = '';
         }
```
Plenty of nested ifs usually make the code more difficult to read, could you extract parts of these into new methods?

AndiBraimllari commented on src/app/text-marking/text-marking.ts:
```diff
@@ -99,10 +111,132 @@ export function markText(
+                addItemAtIndex(markingCount, newTextContent)
+                newNode.style.animation = 'fillBackground 500ms linear forwards';
+            }else{
+                newNode.style.animation = 'fillBackground 500ms linear forwards';
```
Consider using classes instead of assigning CSS as strings in JS. It keeps the CSS code in CSS files.

If you still want to keep it like this, you can create a const for `'fillBackground 500ms linear forwards'` since if I'm not mistaken it is repeated many times.

AndiBraimllari commented on src/app/text-marking/text-marking.ts:
```diff
@@ -99,10 +111,132 @@ export function markText(
+
+      lastHighlightedMarkingWords.splice(index, 0, newItem);
+    } else {
+      console.error('Invalid index');
```
Have you ever encountered a case in which this else case here is hit? The else cases in `removeItemAtIndex` and `addItemAtIndex` feel like they are trying to cover for some silent bugs that might be in `makeTextMarkings`.
That is, why would the indices be out of bounds in these two methods? The answer to this can perhaps be integrated in the `makeTextMarkings` so we can have less uncertainty regarding indices and so on.

AndiBraimllari commented:
I've added some (mostly superficial) comments. Please focus on writing some tests first so we can have some cases to develop and test against.

I will have a deeper look into the logic of this implementation afterwards. Regardless, it seems promising, I simply want to have the code be a bit more rigid.

vaqueraoscar0 commented:
## Details: 
@AndiBraimllari Thanks! I appreciate the feedback.

1. Nested if's - I can definitely break it down a bit further, the only issue i have with the logic is that i will have to move it back to the markText function instead of having it within it's own function, there seems to be something happening when moving the logic within it's own function. Nevertheless, i can reduce and break some stuff down to make it cleaner.

2. The inline styling - I initially had it as a class but ran into a few issues, i've already fixed those issues so i can definitely switch to a class instead (thanks for reminding me).

3. Invalid index - That was mainly for me to test a few things, i simply forgot to remove them. 

4. Test cases - i can definitely add some test cases, the existing test cases might need to be updated to work with the current logic do you want me to update them? the span classes will require to have the new class added.

## final Comments
**Will start working on the changes as soon as possible.**

AndiBraimllari commented:
> ## Details:
> @AndiBraimllari Thanks! I appreciate the feedback.
> 
> 1. Nested if's - I can definitely break it down a bit further, the only issue i have with the logic is that i will have to move it back to the markText function instead of having it within it's own function, there seems to be something happening when moving the logic within it's own function. Nevertheless, i can reduce and break some stuff down to make it cleaner.
> 2. The inline styling - I initially had it as a class but ran into a few issues, i've already fixed those issues so i can definitely switch to a class instead (thanks for reminding me).
> 3. Invalid index - That was mainly for me to test a few things, i simply forgot to remove them.
> 4. Test cases - i can definitely add some test cases, the existing test cases might need to be updated to work with the current logic do you want me to update them? the span classes will require to have the new class added.
> 
> ## final Comments
> **Will start working on the changes as soon as possible.**

Awesome!

1. Only do such refactoring if you genuinely believe that it will improve readability, if you don't see an easy way of doing it please don't over-engineer it preemptively.
2. Alright, if you run into issues again you can keep it in-lined again.
3. Alright.
4. Yes, please update them.

vaqueraoscar0 commented:
@AndiBraimllari I will thanks, and yes the last PR i made, i created some changes to clean the code a bit and didn't test it, i assumed that everything was going to work properly but i was wrong. Am almost done with the changes, i updated the logic and seems to be working a lot better. i am however, still working on the test cases, Am currently running into some issues with the old test cases where some nodes (span elements) may sometimes include the new class and sometimes they don't i believe it may have something to do with a delay, since the result always changes, i added a temporary fix where i remove the animation class before comparing the result. I added a few of my own test cases with the animation class and they seem to be working well am not too sure about the others just yet. Maybe you have some ideas about this. I can commit another PR so you can take a look at this more in depth if you want. I've already done all of the clean up and have improved the logic as well.

AndiBraimllari commented:
Hey, no worries, feel free to push all the code here and I will take a look.

vaqueraoscar0 commented:
@AndiBraimllari Let me know if anything else looks off, i'd be happy to take a look at it and fix it. Regardless, Am still going to be working on this code, and any feedback will help.

AndiBraimllari commented on src/app/text-marking/text-marking.spec.ts:
```diff
@@ -37,6 +37,10 @@ describe('Marker', () => {
         const expectedEditor = document.createElement('div');
         expectedEditor.id = EDITOR_ID;
         expectedEditor.innerHTML = expectedHMTL;
+        editor.innerHTML = editor.innerHTML.replace(
+            /\s*\banimate-marking-text\b\s*/g,
+            ''
+        );
```
If I understand correctly, here you remove all `animate-marking-text` classes?

AndiBraimllari commented on src/app/text-marking/text-marking.ts:
```diff
@@ -101,6 +114,138 @@ export function markText(
+    highlightedMarkingWords.length = 0;
+}
+
+/**
+ * Removes an item from the lastHighlightedMarkingWords array at the specified index.
+ * @param {number} index - The index of the item to be removed from the array.
+ */
+function removeItemAtIndex(index: number): void {
+    lastHighlightedMarkingWords.splice(index, 1);
+}
+
+/**
+ * Adds a new item to the lastHighlightedMarkingWords array at the specified index.
+ * @param {number} index - The index at which to insert the new item.
+ * @param {string} newItem - The new item to be added to the array.
+ */
+function addItemAtIndex(index: number, newItem: string): void {
+    lastHighlightedMarkingWords.splice(index, 0, newItem);
```
I kind of feel like both of these methods can't really justify their existence, what will happen if we will want to use splice with another argument, should we create another method for that? Personally I would just remove them and inline these lines of code where their respective methods are used.

AndiBraimllari commented on src/app/text-marking/text-marking.ts:
```diff
@@ -101,6 +114,138 @@ export function markText(
+ * whether to apply animations to the nodes based on the changes detected.
+ * The logic considers scenarios where nodes are added, removed, or modified.
+ */
+function makeTextMarkings(newNode: HTMLElement, newTextContent: string) {
```
This method is now much more readable! Btw I feel like the terminology can be improved a bit, i.e. can we rename this from `makeTextMarkings` to `applyMarkingFadein`, or something else that is a bit more descriptive.

AndiBraimllari commented:
I'll provide a proper review within 24 hours!

AndiBraimllari commented:
Something that I noticed while running this. It seems that the second marking does not have this fade-in effect applied. It could be due to the fact that `markText` is called for each paragraph, https://github.com/OpenCovenant/quill/blob/main/src/app/home/home.component.ts#L248 and that the state that you keep in the variables in the `text-marking.ts` file is kept across `markText` calls and that in turn might cause an issue.

Can you confirm that this is a bug?

[Screencast from 10-04-2023 10:23:21 PM.webm](https://github.com/OpenCovenant/quill/assets/11479070/0ca43216-0ee0-4f2b-84fa-c473a377f7f5)

AndiBraimllari commented:
Please write a test about this, and push it even if it fails.

AndiBraimllari commented:
You can store such variables (e.g. `highlightedMarkingWords`, `lastHighlightedMarkingWords`, ...) in https://github.com/OpenCovenant/quill/blob/main/src/app/home/home.component.ts#L248, so that they pertain to each paragraph call, and they are available in the `markText` method as arguments, and then you might also not need methods such as `isLastCall` because you'll know when `markText` has finished. Let me know if I'm overlooking something

vaqueraoscar0 commented:
@AndiBraimllari Yes am currently working on that issue. The problem i have here is that with a new line involved there seems to be a seperate p element created containing a different set of typo's, what i need here is the total number of marked text but because the markedText function thats being used from home.component calls this multiple times for every new p element am unable to get the exact total number.

AndiBraimllari commented:
Yup, see my last comment, perhaps you can try storing such code in the home component right before or during each paragraph call to `markText`. I assume you should have access to what you need there?

By the way, writing such code is not easy, writing the `markText` method took me some effort, you've already done an impressive job so far :)

If you want me to spend more time with this bug, please let me know and I'll do so.

vaqueraoscar0 commented:
@AndiBraimllari The isLastcall that i created is based on the total of number of textMarkings but once again because the function is being called multiple times whenever a new line is created am only able to get the total number of typo's from the elements that were from one p. i was trying really hard to avoid using any other components but yeah it would be a lot easier if i can also modify the home.component.ts file

AndiBraimllari commented:
Feel free to make changes anywhere you see fit, no reason for me to stop you from writing a better implementation.

vaqueraoscar0 commented:
@AndiBraimllari Thanks i will, that's going to make it a bit easier for me to work on this.

vaqueraoscar0 commented:
@AndiBraimllari Hey sorry it's taking so long, i've already fixed the newline issue, but found yet another thing that needs to be fixed. Am not really sure if your aware or maybe am missing something but there seem to be an issue related to duplicated typo's and spaces where the  hightlighted text for both the space and the duplication seems to disappear after a second call. I though it had something to do with my code but after testing it, i realized it was something outside my own code. 

Note: in the video am running the program without my code.


https://github.com/OpenCovenant/quill/assets/57508905/629fade0-7fc5-47ee-aa4e-afb920afb692

AndiBraimllari commented:
Hey, first off, no worries, we don't really have any urgency :)
I'll try to break some things down in a comment below.

AndiBraimllari commented:
You'll notice that we support markings that overlap (imagine an entire sentence marked in blue, and it has a typo, the red and the green marking would both be shown). This is the case for the "asd" marking. The first marking is about consecutively written words, which are usually a writing mistake by the user (this happens basically in any language, in English think of "He was was jogging"), Additionally, the repeated word is a typo, therefore we see such markings as in the screenshot.
![Screenshot from 2023-10-05 23-41-49](https://github.com/OpenCovenant/quill/assets/11479070/e1e90462-8b0d-47a5-bda2-8c0665dc69cf)

AndiBraimllari commented:
Regarding the marking for too many spaces, this one is a bit tricky. To explain this, think of markings for sentences and markings for words on their own. When it comes to processing, the period in a sentence is not necessarily what marks the ending of a sentence, as it can be used for abbreviations (think in English of "He was writing i.e. he was using a pencil.")
The "asd asd" you add in the end of the video seem to throw the processing of the sentence a bit off, and therefore ignoring the multiple spaces marking, which I guess we should improve.

AndiBraimllari commented:
If you believe that any of this prohibits you in implementing this feature, then please let me know (or simply create a ticket and we will highly prioritize it).

vaqueraoscar0 commented:
@AndiBraimllari Oh i see, but wouldn't the typo still need to be listed on the side and not be taken out, so for example on the first run we had 'asd asd', 'asd', 'asd', ' ', listed and on the second run only 'asd', 'asd', remain, and both the space and the duplicate get taken of the list. Either way if that's normal, then i might have a way to work around it to get my code working properly.

AndiBraimllari commented:
Both markings (multiple subsequent spaces and consecutively repeated word) are implemented at the sentence level (because you need more “context” than just each word in itself, and a sentence basically is defined here if it ends with a comma AND the next word starts with a capital letter (otherwise it could be an abbreviation, hard to tell).

in reality this wouldn’t really be an issue because sentences usually end with a period and thr next letter after that is a capital letter (in your video try writing “Asd asd” in the end instead). I don’t think your code should really depend on such a behavior.

As always, frel free to ping me, happy to help :)

vaqueraoscar0 commented:
@AndiBraimllari Awesome! I appreciate the explanation. And i will thanks.

AndiBraimllari commented:
Do you plan to make further changes here? Is it ready to be reviewed again?

vaqueraoscar0 commented:
@AndiBraimllari Yes i believe it should be ready now, Everything seems to run smooth with this new implementation, not a whole lot was changed, i tried making some big changes inside of the home.component but unfortunately, it didn't work out as i plan so i decided to keep it somewhat the same. I tested this for a while now and haven't come up with any bugs except one, but i haven't been able to replicate the bug again. The bug if i remember correctly happen because the number of text markings increased and became greater then the number of words listed, i had printed to the console and saw that the consumableTextMarkings had somehow identified the normal text as typo's, so there was a mixture of both typos and regular text being listing inside of the array. Let me know if there's anything else you want me to do here. Or if you know anything about the bug i mentioned earlier.
